<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Paul Staab" />

<meta name="date" content="2024-01-08" />

<title>Command Line Arguments</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Command Line Arguments</h1>
<h4 class="author">Paul Staab</h4>
<h4 class="date">2024-01-08</h4>



<p>scrm is uses a syntax compatible with the popular program <a href="http://home.uchicago.edu/~rhudson1/source/mksamples.html">ms</a>.
There are, however, a few differences to ms:</p>
<ul>
<li>scrm can not simulate
<ul>
<li>gene conversions (<code>-c</code> in ms) and</li>
<li>a fix number of segregating sites (<code>-s</code>),</li>
</ul></li>
<li>the option <code>-L</code> produces a slightly different output
and</li>
<li>it additionally implements the flags
<ul>
<li><code>-l</code> (approximation),</li>
<li><code>-sr</code> (changing recombination rate),</li>
<li><code>-st</code> (changing mutation rate),</li>
<li><code>-eI</code> (sampling haplotypes at multiple time points)
and</li>
<li><code>-oSFS</code> (generates frequency spectra).</li>
</ul></li>
<li>We do not support changing the number of populations with
<code>-ema</code>. Our version of the command is just
<code>-ema &lt;t&gt; &lt;M11&gt; &lt;M12&gt; ...</code> instead of
<code>-ema &lt;t&gt; &lt;npop&gt; &lt;M11&gt; &lt;M12&gt; ...</code>.</li>
</ul>
<p>For all other options, you can also refer to <a href="http://home.uchicago.edu/~rhudson1/source/mksamples.html">ms’
manual</a> to get a detailed description of what the commands are doing.
scrm should happily execute any ms command that does not contain
<code>-c</code>, <code>-s</code> and <code>-ema</code>. Also, scrm has
somewhat stricter requirements regarding the order of arguments if
population admixture (<code>-es</code>) is involved.</p>
<div id="general-syntax" class="section level2">
<h2>General Syntax</h2>
<p>The arguments for calling <em>scrm</em> are</p>
<pre><code>scrm &lt;nhap&gt; &lt;nrep&gt; [...]</code></pre>
<p>where <em>nhap</em> is the total number of haplotypes (in all
populations and at all times) that are simulated at each locus, and
<em>nrep</em> is the number of independent loci that will be produced.
The <code>[...]</code> is an optional placeholder for an arbitrary
number of command line flags described below.</p>
</div>
<div id="basic-parameters" class="section level2">
<h2>Basic Parameters</h2>
<div id="recombination" class="section level3">
<h3>Recombination</h3>
<ul>
<li><code>-r &lt;R&gt; &lt;L&gt;</code>: Set the recombination rate to
<em>R = 4<em>N0</em>r</em> and the length of all loci to L base pairs.
<em>r</em> is expected number of recombinations on the locus per
generation.</li>
<li><code>-l &lt;l&gt;</code>: Use approximation rather than simulating
the exact ARG. Within a sliding window of length <em>l</em> base pairs
all linkage information is considered when building the genealogy. To
positions outside of this window, some linkage is ignored. Setting
<em>l=0</em> produces the SMC’ and <em>l=-1</em> deactivates the
approximation. Since v1.6.0, it’s also possible to specify the window’s
length in number of recombinations. To do so, use
<code>-l &lt;x&gt;r</code>, where x is the number of recombinations
(e.g. <code>-l 100r</code> for a window spanning 100 recombinations).
Also starting with version 1.6.0 <strong>approximation is turned on by
default</strong> using a conservative window length of 500
recombinations. For most applications, it should be fine to reduce this
value to 100 - 250 recombinations if runtime is a critical factor.</li>
</ul>
</div>
<div id="population-structure-migration" class="section level3">
<h3>Population structure &amp; migration</h3>
<p>In all commands, migrations rates <em>M = 4<em>N0</em>m</em>, where
<em>m</em> is the fraction of a population that is replaced with
migrants from other populations each generation (looking forwards in
time).</p>
<ul>
<li><code>-I &lt;npop&gt; &lt;s1&gt; ... &lt;sn&gt; [&lt;M&gt;]</code>:
Use an island model with <em>npop</em> populations, where <em>s1</em> to
<em>sn</em> haplotypes are sampled from population 1 to n, respectively.
Optionally assume a symmetric migration rate of <em>M</em>.</li>
<li><code>-M &lt;M&gt;</code>: Assume a symmetric migration rate of
<em>M/(npop-1)</em>.</li>
<li><code>-m &lt;i&gt; &lt;j&gt; &lt;M&gt;</code>: Set the migration
rate from population <em>j</em> to population <em>i</em> to <em>M</em>
(looking forward in time) [since v1.3.1].</li>
<li><code>-ma &lt;M11&gt; &lt;M21&gt; ... &lt;M21&gt; ...</code>: Set
the migration matrix (Dimension is <em>npop x npop</em>). Diagonals
elements are ignored but required (you can use <code>x</code> or
<code>0</code>).</li>
</ul>
</div>
<div id="population-size-changes" class="section level3">
<h3>Population size changes</h3>
<p>For exponential growth/decline of a population, the parameter
<em>a</em> changes the size of a population according to the formula
<em>N(s) = N(0)<em>exp(-a</em>s)</em>, where <em>N(0)</em> is the
population’s size at the time of the command (e.g. <em>0</em> for
<code>-g &lt;a&gt;</code> and <code>-G &lt;a&gt;</code> and <em>t</em>
for <code>-eg &lt;t&gt; &lt;a&gt;</code> and
<code>-eG &lt;t&gt; &lt;a&gt;</code>) and <em>N(s)</em> is the size of
the population <em>s</em> time units in the past. Looking forwards in
time, a positive <em>a</em> leads to population growth, while a negative
one generates a decline in population sizes.</p>
<ul>
<li><code>-n &lt;i&gt; &lt;n&gt;</code>: Set the present day size of
population <em>i</em> to _n*N0_.</li>
<li><code>-G &lt;a&gt;</code>: Set the exponential growth rate of all
populations to <em>a</em>.</li>
<li><code>-g &lt;i&gt; &lt;a&gt;</code>: Set the exponential growth rate
of population <em>i</em> to <em>a</em>.</li>
</ul>
</div>
<div id="summary-statistics" class="section level3">
<h3>Summary Statistics</h3>
<ul>
<li><code>-t &lt; $\theta$ &gt;</code>: Set the mutation rate to <span class="math inline">\(\theta = 4N_0u\)</span>, where <em>u</em> is the
neutral mutation rate per locus. If this options is given, scrm
generates the segregating sites output.</li>
<li><code>-transpose-segsites</code> or
<code>--transpose-segsites</code>: If given, the segregating sites are
printed with each row representing a mutation and each column
representing a haplotype, rather than the other way round. Additionally,
the time at which a mutation occurred is reported (in units of <em>4 *
N0</em> generations) [since v1.7.0].</li>
<li><code>-T</code>: Print the local genealogies in newick format.</li>
<li><code>-O</code>: Print the local genealogies in the
<code>oriented forest</code> format as described in <a href="https://www.sciencedirect.com/science/article/pii/S0040580914000355">Kelleher
<em>et al.</em> (2014)</a> [since v1.2].</li>
<li><code>-L</code>: Print the TMRCA and the local tree length for each
segment (behaves different to ms). Both values are scaled in coalescent
time units, e.g. in <em>4 * N0</em> generations.</li>
<li><code>-oSFS</code>: Print the site frequency spectrum. Requires that
the mutation rate <span class="math inline">\(\theta\)</span> is given
with the ‘-t’ option.</li>
<li><code>-SC [ms|rel|abs]</code>: Scaling of sequence positions. Either
relative to the locus length between 0 and 1 (<code>rel</code>),
absolute in base pairs (<code>abs</code>) or <code>ms</code>’s scaling
(default) where the positions in the <em>segregating sites</em> output
are relative, and the positions in the trees output are absolute
(<code>ms</code>) [since v1.3.0].</li>
</ul>
</div>
<div id="other-options" class="section level3">
<h3>Other options</h3>
<ul>
<li><code>-seed &lt;SEED&gt; [&lt;SEED2&gt; &lt;SEED3&gt;]</code>:
Specifies a seed for the simulation. You can input up to three
non-negative numbers. If no seed is given, scrm generates one using
entropy provided by the operating system. To reproduce a previous
simulation, use the single number in the second line of the output.</li>
<li><code>-print-model, --print-model</code>: Prints information about
the model defined by the command line arguments, including calculated
population sizes. Can be useful for debugging or verifying the model
[since v1.5.0].</li>
<li><code>-p &lt;digits&gt;</code>: Sets the number of significant
digits used in the output [since v1.4.0].</li>
<li><code>-h</code>, <code>--help</code>: Prints a help text.</li>
<li><code>-v</code>, <code>--version</code>: Prints version
information.</li>
</ul>
</div>
</div>
<div id="time-specific-parameters" class="section level2">
<h2>Time specific parameters</h2>
<p>The command this section all have a time <em>t</em> as first
parameter. Changes made by the commands affect the time from <em>t</em>
further back into the past. All times in units of _4*N0_
generations.</p>
<div id="population-structure-migration-1" class="section level3">
<h3>Population structure &amp; migration</h3>
<ul>
<li><code>-eI &lt;t&gt; &lt;s1&gt; ... &lt;sn&gt;</code>: Sample
<em>s1</em> to <em>sn</em> haplotypes are from population <em>1</em> to
<em>n</em>, respectively, at time <em>t</em>.</li>
<li><code>-eM &lt;t&gt; &lt;M&gt;</code>: Assume a symmetric migration
rate of <em>M/(npop-1)</em> at time <em>t</em>.</li>
<li><code>-em &lt;t&gt; &lt;i&gt; &lt;j&gt; &lt;M&gt;</code>: Set the
migration rate from population <em>j</em> to population <em>i</em> to
<em>M</em> (looking forward in time) at time <em>t</em> [since
v1.3.1].</li>
<li><code>-ema &lt;t&gt; &lt;M11&gt; &lt;M12&gt; ... &lt;M21&gt; ...</code>:
Set the migration matrix at time <em>t</em> (Dimension is <em>npop x
npop</em>). Diagonals elements are ignored but required (use ‘x’ or 0).
The rates apply pastwards from time <em>t</em>.</li>
</ul>
</div>
<div id="population-size-changes-1" class="section level3">
<h3>Population size changes</h3>
<ul>
<li><code>-eN &lt;t&gt; &lt;n&gt;</code>: Set the size of all
populations to _n*N0_ at time <em>t</em>.</li>
<li><code>-en &lt;t&gt; &lt;i&gt; &lt;n&gt;</code>: Set the size of
population <em>i</em> to _n*N0_ at time <em>t</em>.</li>
<li><code>-eg &lt;t&gt; &lt;i&gt; &lt;a&gt;</code>: Set the exponential
growth rate of population <em>i</em> to <em>a</em> at time
<em>t</em>.</li>
<li><code>-eG &lt;t&gt; &lt;a&gt;</code>: Set the exponential growth
rate of all populations to <em>a</em> at time <em>t</em>.</li>
</ul>
</div>
<div id="population-splits-merges" class="section level3">
<h3>Population Splits &amp; Merges</h3>
<ul>
<li><code>-es &lt;t&gt; &lt;i&gt; &lt;p&gt;</code>: Population
admixture. Replaces a fraction <em>1-p</em> of population <em>i</em>
with haplotypes from a population <em>npop + 1</em>. Technically (and
looking backwards in time), a new population <em>n+1</em> with size
<em>N0</em> is created at time <em>t</em>. Migration (to &amp; from) and
growth rates for this population are initially 0. Each lines in
population <em>i</em> is moved to the new population with probability
<em>1-p</em>. Please sort multiple <code>-es</code> arguments by their
time to avoid confusion about the numbering of populations. Please give
the arguments that affect the whole population (<code>-M</code>,
<code>-N</code>, <code>-G</code> &amp; <code>-ma</code>) before giving
the first <code>-es</code>. Also, their timed equivalent’s
(<code>-eM</code>, <code>-eN</code>, <code>-eG</code>, <code>-eI</code>
&amp; <code>-ema</code>) position on the command line events must also
be sorted by time, at least relative to the <code>-es</code> argument.
<code>scrm</code> throws an error if any of these conditions is not met.
In doubt, just sort all command line arguments by their time.</li>
<li><code>-eps &lt;t&gt; &lt;i&gt; &lt;j&gt; &lt;p&gt;</code>: Partial
admixture. Similar to <code>-es</code> but replaces a fraction
<code>1-p</code> of population <em>i</em> with haploids from population
<em>j</em> at time <em>t</em>. Different to <code>-es</code>, population
<em>j</em> is a normal population that continues to exist at times more
recent than <em>t</em>. Viewed backwards in time, this moves a fraction
<em>1-p</em> of the linages in population <em>i</em> to population
<em>j</em>. This does not change the number of populations, population
sizes, growth or migration rates in any way [since v1.5.0].</li>
<li><code>-ej &lt;t&gt; &lt;j&gt; &lt;i&gt;</code>: Adds a
specialization event in population <em>i</em> that creates population
<em>j</em> (forwards in time). Technically (and looking backwards in
time), it moves all lines from population <em>j</em> into population
<em>i</em> at time <em>t</em>. Migration rates into population
<em>j</em> are set to 0 for the time further back into the past.</li>
</ul>
<p>When multiple <code>es</code>, <code>eps</code> or <code>ej</code>
arguments are given for the same time <em>t</em>, the migrations are
executed in the order in which the commands are given. For example if we
have <code>-es 0.08 2 .2 -ej 0.08 3 1</code>, first 80% of pop 2 move to
a newly created pop 3 (viewed backwards in time), then everyone that
just moved to pop 3 moves on to pop 1. This is equivalent to
<code>-eps 0.08 2 1 .2</code>, except that the latter does not create
the empty population 3.</p>
</div>
</div>
<div id="sequence-specific-parameters" class="section level2">
<h2>Sequence specific parameters</h2>
<p>The following commands change the model parameters from at a sequence
position <em>s</em>. You should still set the initial rate with
<code>-r</code> or <code>-t</code>, respectively, and then use the
commands prefixed with <code>s</code> for all changes. Note that
<code>-r</code> also takes the total length of the sequence as second
argument, while <code>-sr</code> just has the rate as argument.</p>
<ul>
<li><code>-sr &lt;s&gt; &lt;R&gt;</code>: Set the recombination rate to
<em>R</em> starting at position <em>s</em>.</li>
<li><code>-st &lt;s&gt; &lt;$\theta$&gt;</code>: Set the mutation rate
to <span class="math inline">\(\theta\)</span> starting at position
<em>s</em>.</li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
